core/utils/historical-data/src/lib.rs

```diff
@@ -0,0 +1,59 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! History driven data storage.
+//! Useful to store information with history
+//! on a per item basis.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use rstd::convert::TryInto;
+
+pub mod tree;
+pub mod linear;
+
+/// An entry at a given history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+pub struct HistoricalValue<V, I> {
+	/// The stored value.
+	pub value: V,
+	/// The moment in history when the value got set.
+	pub index: I,
+}
+
+impl<V, I> From<(V, I)> for HistoricalValue<V, I> {
+	fn from(input: (V, I)) -> HistoricalValue<V, I> {
+		HistoricalValue { value: input.0, index: input.1 }
+	}
+}
```

Comment:
Usually I prefer when From/Into are used to convert between types, here `(V, I)` is not really a type, implementing a function `new` seems to make more sense here to me.